extensions/sql/data/WriteCallback.h

```diff
@@ -0,0 +1,49 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <string>
+
+#include "FlowFileRecord.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+
+class WriteCallback : public OutputStreamCallback {
+public:
+  WriteCallback(const std::string& data)
```

Comment:
Single parameter constructors should be marked as `explicit` or have a code comment explaining why implicit conversion is desired from the parameter type.

See also my reply to the thread about `const_cast`. If we mutate through the pointer/object, then it shouldn't be `const`. The caller should be responsible for creating a mutable buffer, potentially by casting away `const` from their `str.data()` or any other way.

I also prefer the old signature (`char*` + length) because it only requires mutable access from the caller, whereas `std::string` requires ownership as well, since `std::string` is an owning container.

Reply:
Signature could be WriteCallback(std::string data) and then for optimization caller can call `WriteCallback writer(std::move(output));` and in this particular case it works, but it won't if `output` is used later in the code.

`data` is not changed and because of this `WriteCallback(const std::string& data)` has const reference. We need to remove const because `WriteCallback` uses  `DataStream::writeData(uint8_t *value, int size);` (which should be `const uint8_t* value` BTW), but this is internals of WriteCallback, and it's callers should not know about it.

Reply:
I've misunderstood the purpose of the class. Up until now I thought that `stream->write` writes into `data_`, therefore modifies it. I only now realized that it copies _from_ `data_` _to_ the stream, not the other way around, and that the incorrect interface here is `BaseStream::write`, not `std::string::data`.

Sorry for my misunderstanding, feel free to close this.